### ISSUE TITLE
False positive for CVE-2021-24031 on Ubuntu 20.04

### ISSUE THREAD

**Issue by ginkel** [aquasecurity/trivy#912]
## Description

When scanning the `ubuntu:20.04` Docker image using `trivy image`, the image is reported as being vulnerable to CVE-2021-24031 despite version `1.4.4+dfsg-3ubuntu0.1` of the `libzstd1` package being installed and classified as "Fixed" by Ubuntu, cf. https://ubuntu.com/security/CVE-2021-24031.

## What did you expect to happen?

CVE-2021-24031 is detected as fixed.

## What happened instead?

The image is detected as vulnerable to CVE-2021-24031.

## Output of run with `-debug`:

```
$ trivy --debug i ubuntu:20.04
2021-03-23T10:34:32.265+0100    DEBUG   Severities: UNKNOWN,LOW,MEDIUM,HIGH,CRITICAL
2021-03-23T10:34:32.278+0100    DEBUG   cache dir:  /home/tg/.cache/trivy
2021-03-23T10:34:32.278+0100    DEBUG   DB update was skipped because DB is the latest
2021-03-23T10:34:32.278+0100    DEBUG   DB Schema: 1, Type: 1, UpdatedAt: 2021-03-23 00:36:29.688307682 +0000 UTC, NextUpdate: 2021-03-23 12:36:29.688307382 +0000 UTC, DownloadedAt: 2021-03-23 09:03:43.16886849 +0000 UTC
2021-03-23T10:34:32.280+0100    DEBUG   Vulnerability type:  [os library]
2021-03-23T10:34:32.280+0100    DEBUG   Artifact ID: sha256:4dd97cefde62cf2d6bcfd8f2c0300a24fbcddbe0ebcd577cc8b420c29106869a
2021-03-23T10:34:32.280+0100    DEBUG   Blob IDs: [sha256:aeb3f02e937406fb402a926ce5cebc7da79b14dbcb4f85a5ce0e3855623cec80 sha256:db978cae6a05486620211667eea3a3174eed3afe6e2a21a1d00e3f05c459a72e sha256:c20d459170d8e39bbf5b9ef390136ffbe894caac7823116466008c02d38836ac]
2021-03-23T10:34:32.281+0100    INFO    Detecting Ubuntu vulnerabilities...
2021-03-23T10:34:32.281+0100    DEBUG   ubuntu: os version: 20.04
2021-03-23T10:34:32.281+0100    DEBUG   ubuntu: the number of packages: 92
2021-03-23T10:34:32.281+0100    INFO    Trivy skips scanning programming language libraries because no supported file was detected

ubuntu:20.04 (ubuntu 20.04)
===========================
Total: 26 (UNKNOWN: 0, LOW: 22, MEDIUM: 4, HIGH: 0, CRITICAL: 0)

+-------------+------------------+----------+------------------------+-----------------------+-----------------------------------------+
|   LIBRARY   | VULNERABILITY ID | SEVERITY |   INSTALLED VERSION    |     FIXED VERSION     |                  TITLE                  |
+-------------+------------------+----------+------------------------+-----------------------+-----------------------------------------+
| bash        | CVE-2019-18276   | LOW      | 5.0-6ubuntu1.1         |                       | bash: when effective UID is not         |
|             |                  |          |                        |                       | equal to its real UID the...            |
|             |                  |          |                        |                       | -->avd.aquasec.com/nvd/cve-2019-18276   |
+-------------+------------------+          +------------------------+-----------------------+-----------------------------------------+
| coreutils   | CVE-2016-2781    |          | 8.30-3ubuntu2          |                       | coreutils: Non-privileged               |
|             |                  |          |                        |                       | session can escape to the               |
|             |                  |          |                        |                       | parent session in chroot                |
|             |                  |          |                        |                       | -->avd.aquasec.com/nvd/cve-2016-2781    |
+-------------+------------------+          +------------------------+-----------------------+-----------------------------------------+
| gpgv        | CVE-2019-13050   |          | 2.2.19-3ubuntu2.1      |                       | GnuPG: interaction between the          |
|             |                  |          |                        |                       | sks-keyserver code and GnuPG            |
|             |                  |          |                        |                       | allows for a Certificate...             |
|             |                  |          |                        |                       | -->avd.aquasec.com/nvd/cve-2019-13050   |
+-------------+------------------+          +------------------------+-----------------------+-----------------------------------------+
| libc-bin    | CVE-2016-10228   |          | 2.31-0ubuntu9.2        |                       | glibc: iconv program can hang           |
|             |                  |          |                        |                       | when invoked with the -c option         |
|             |                  |          |                        |                       | -->avd.aquasec.com/nvd/cve-2016-10228   |
+             +------------------+          +                        +-----------------------+-----------------------------------------+
|             | CVE-2019-25013   |          |                        |                       | glibc: buffer over-read in              |
|             |                  |          |                        |                       | iconv when processing invalid           |
|             |                  |          |                        |                       | multi-byte input sequences in...        |
|             |                  |          |                        |                       | -->avd.aquasec.com/nvd/cve-2019-25013   |
+             +------------------+          +                        +-----------------------+-----------------------------------------+
|             | CVE-2020-27618   |          |                        |                       | glibc: iconv when processing            |
|             |                  |          |                        |                       | invalid multi-byte input                |
|             |                  |          |                        |                       | sequences fails to advance the...       |
|             |                  |          |                        |                       | -->avd.aquasec.com/nvd/cve-2020-27618   |
+             +------------------+          +                        +-----------------------+-----------------------------------------+
|             | CVE-2020-29562   |          |                        |                       | glibc: assertion failure in iconv       |
|             |                  |          |                        |                       | when converting invalid UCS4            |
|             |                  |          |                        |                       | -->avd.aquasec.com/nvd/cve-2020-29562   |
+             +------------------+          +                        +-----------------------+-----------------------------------------+
|             | CVE-2020-6096    |          |                        |                       | glibc: signed comparison                |
|             |                  |          |                        |                       | vulnerability in the                    |
|             |                  |          |                        |                       | ARMv7 memcpy function                   |
|             |                  |          |                        |                       | -->avd.aquasec.com/nvd/cve-2020-6096    |
+-------------+------------------+          +                        +-----------------------+-----------------------------------------+
| libc6       | CVE-2016-10228   |          |                        |                       | glibc: iconv program can hang           |
|             |                  |          |                        |                       | when invoked with the -c option         |
|             |                  |          |                        |                       | -->avd.aquasec.com/nvd/cve-2016-10228   |
+             +------------------+          +                        +-----------------------+-----------------------------------------+
|             | CVE-2019-25013   |          |                        |                       | glibc: buffer over-read in              |
|             |                  |          |                        |                       | iconv when processing invalid           |
|             |                  |          |                        |                       | multi-byte input sequences in...        |
|             |                  |          |                        |                       | -->avd.aquasec.com/nvd/cve-2019-25013   |
+             +------------------+          +                        +-----------------------+-----------------------------------------+
|             | CVE-2020-27618   |          |                        |                       | glibc: iconv when processing            |
|             |                  |          |                        |                       | invalid multi-byte input                |
|             |                  |          |                        |                       | sequences fails to advance the...       |
|             |                  |          |                        |                       | -->avd.aquasec.com/nvd/cve-2020-27618   |
+             +------------------+          +                        +-----------------------+-----------------------------------------+
|             | CVE-2020-29562   |          |                        |                       | glibc: assertion failure in iconv       |
|             |                  |          |                        |                       | when converting invalid UCS4            |
|             |                  |          |                        |                       | -->avd.aquasec.com/nvd/cve-2020-29562   |
+             +------------------+          +                        +-----------------------+-----------------------------------------+
|             | CVE-2020-6096    |          |                        |                       | glibc: signed comparison                |
|             |                  |          |                        |                       | vulnerability in the                    |
|             |                  |          |                        |                       | ARMv7 memcpy function                   |
|             |                  |          |                        |                       | -->avd.aquasec.com/nvd/cve-2020-6096    |
+-------------+------------------+          +------------------------+-----------------------+-----------------------------------------+
| libgcrypt20 | CVE-2019-12904   |          | 1.8.5-5ubuntu1         |                       | Libgcrypt: physical addresses           |
|             |                  |          |                        |                       | being available to other processes      |
|             |                  |          |                        |                       | leads to a flush-and-reload...          |
|             |                  |          |                        |                       | -->avd.aquasec.com/nvd/cve-2019-12904   |
+-------------+------------------+          +------------------------+-----------------------+-----------------------------------------+
| libgnutls30 | CVE-2021-20231   |          | 3.6.13-2ubuntu1.3      |                       | gnutls: Use after free in               |
|             |                  |          |                        |                       | client key_share extension              |
|             |                  |          |                        |                       | -->avd.aquasec.com/nvd/cve-2021-20231   |
+             +------------------+          +                        +-----------------------+-----------------------------------------+
|             | CVE-2021-20232   |          |                        |                       | gnutls: Use after free                  |
|             |                  |          |                        |                       | in client_send_params in                |
|             |                  |          |                        |                       | lib/ext/pre_shared_key.c                |
|             |                  |          |                        |                       | -->avd.aquasec.com/nvd/cve-2021-20232   |
+-------------+------------------+          +------------------------+-----------------------+-----------------------------------------+
| libpcre3    | CVE-2017-11164   |          | 2:8.39-12build1        |                       | pcre: OP_KETRMAX feature in the         |
|             |                  |          |                        |                       | match function in pcre_exec.c           |
|             |                  |          |                        |                       | -->avd.aquasec.com/nvd/cve-2017-11164   |
+             +------------------+          +                        +-----------------------+-----------------------------------------+
|             | CVE-2019-20838   |          |                        |                       | pcre: buffer over-read in               |
|             |                  |          |                        |                       | JIT when UTF is disabled                |
|             |                  |          |                        |                       | -->avd.aquasec.com/nvd/cve-2019-20838   |
+             +------------------+          +                        +-----------------------+-----------------------------------------+
|             | CVE-2020-14155   |          |                        |                       | pcre: integer overflow in libpcre       |
|             |                  |          |                        |                       | -->avd.aquasec.com/nvd/cve-2020-14155   |
+-------------+------------------+----------+------------------------+-----------------------+-----------------------------------------+
| libsystemd0 | CVE-2018-20839   | MEDIUM   | 245.4-4ubuntu3.4       |                       | systemd: mishandling of the             |
|             |                  |          |                        |                       | current keyboard mode check             |
|             |                  |          |                        |                       | leading to passwords being...           |
|             |                  |          |                        |                       | -->avd.aquasec.com/nvd/cve-2018-20839   |
+-------------+------------------+----------+------------------------+-----------------------+-----------------------------------------+
| libtasn1-6  | CVE-2018-1000654 | LOW      | 4.16.0-2               |                       | libtasn1: Infinite loop in              |
|             |                  |          |                        |                       | _asn1_expand_object_id(ptree)           |
|             |                  |          |                        |                       | leads to memory exhaustion              |
|             |                  |          |                        |                       | -->avd.aquasec.com/nvd/cve-2018-1000654 |
+-------------+------------------+----------+------------------------+-----------------------+-----------------------------------------+
| libudev1    | CVE-2018-20839   | MEDIUM   | 245.4-4ubuntu3.4       |                       | systemd: mishandling of the             |
|             |                  |          |                        |                       | current keyboard mode check             |
|             |                  |          |                        |                       | leading to passwords being...           |
|             |                  |          |                        |                       | -->avd.aquasec.com/nvd/cve-2018-20839   |
+-------------+------------------+          +------------------------+-----------------------+-----------------------------------------+
| libzstd1    | CVE-2021-24031   |          | 1.4.4+dfsg-3           | 1.4.4+dfsg-3ubuntu0.1 | zstd: adds read permissions             |
|             |                  |          |                        |                       | to files while being                    |
|             |                  |          |                        |                       | compressed or uncompressed              |
|             |                  |          |                        |                       | -->avd.aquasec.com/nvd/cve-2021-24031   |
+             +------------------+          +                        +                       +-----------------------------------------+
|             | CVE-2021-24032   |          |                        |                       | zstd: Race condition                    |
|             |                  |          |                        |                       | allows attacker to access               |
|             |                  |          |                        |                       | world-readable destination file         |
|             |                  |          |                        |                       | -->avd.aquasec.com/nvd/cve-2021-24032   |
+-------------+------------------+----------+------------------------+-----------------------+-----------------------------------------+
| login       | CVE-2013-4235    | LOW      | 1:4.8.1-1ubuntu5.20.04 |                       | shadow-utils: TOCTOU race               |
|             |                  |          |                        |                       | conditions by copying and               |
|             |                  |          |                        |                       | removing directory trees                |
|             |                  |          |                        |                       | -->avd.aquasec.com/nvd/cve-2013-4235    |
+-------------+                  +          +                        +-----------------------+                                         +
| passwd      |                  |          |                        |                       |                                         |
|             |                  |          |                        |                       |                                         |
|             |                  |          |                        |                       |                                         |
|             |                  |          |                        |                       |                                         |
+-------------+------------------+----------+------------------------+-----------------------+-----------------------------------------+
```

## Output of `trivy -v`:

```
$ trivy -v
Version: 0.16.0
Vulnerability DB:
  Type: Light
  Version: 1
  UpdatedAt: 2021-03-23 00:36:29.688307682 +0000 UTC
  NextUpdate: 2021-03-23 12:36:29.688307382 +0000 UTC
  DownloadedAt: 2021-03-23 09:03:43.16886849 +0000 UTC
```

**Comment by ronaudinho** [comment]
```
$ docker run --rm -it ubuntu:20.04 bash
root@7813822ae625:/# dpkg -l | grep zstd
ii  libzstd1:amd64          1.4.4+dfsg-3               amd64        fast lossless compression algorithm
```
by default, `1.4.4+dfsg-3` is the installed version in `ubuntu:20.04` docker image, which is vulnerable
```
root@63ece0fb8b45:/# apt update
root@63ece0fb8b45:/# apt list --upgradable | grep zstd
WARNING: apt does not have a stable CLI interface. Use with caution in scripts.
libzstd1/focal-updates,focal-security 1.4.4+dfsg-3ubuntu0.1 amd64 [upgradable from: 1.4.4+dfsg-3]
```
running `apt update` it is required to get the fixed  `1.4.4+dfsg-3ubuntu0.1` version

**Comment by knqyf263** [comment]
@ronaudinho Thank you for debugging.
@ginkel `ubuntu:20.04` has libzstd1 1.4.4+dfsg-3 and it is vulnerable. Feel free to reopen the issue if you have a further question.
